eventually-v0-5-0/src/version.rs

```diff
@@ -0,0 +1,52 @@
+use std::{
+    fmt::{Debug, Display, Formatter, Result as FmtResult},
+    ops::Add,
+};
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Version(pub u64);
```

Comment:
what does the newtype give you that a naked u64 doesn't? (especially since making the inner type `pub` breaks any encapsulation the pattern might give you)

Reply:
I see your point, I've struggled with this one myself a bit, unsure whether this approach made sense.

On one hand I like the explicitness of `Version` -- as a type name -- over `u64` (this goes towards the DDD approach of naming types/value-objects explicitly, even if they are simple primitives). I also like how it restricts the amount operations allowed on those types. Also it makes it easier to refactor this type if, say in the future, we might want to change the underlying primitive type for this.

However I understand how using `pub` defeats any kind of encapsulation 😅 
I still like the explicitness in what the type name gives though, and I don't like the tediousness a `From<u64>` introduces (which would be `Version::from(0)` or worse `0.into()`)

Reply:
> On one hand I like the explicitness of Version -- as a type name -- over u64 (this goes towards the DDD approach of naming types/value-objects explicitly, even if they are simple primitives).

maybe a type alias is more appropriate here than a newtype. Personally I think both are over-kill. you need to carefully weigh the benefits of an explicit DSL, over the extra cognitive load that DSL introduces for a consumer of this library. If it looks like a u64, acts like a u64, and smells like a u64, why not use a u64?


>I also like how it restricts the amount operations allowed on those types. Also it makes it easier to refactor this type if, say in the future, we might want to change the underlying primitive type for this.
> 
> However I understand how using pub defeats any kind of encapsulation 😅

ha exactly. since it's pub, that refactor would be a breaking change anyway

> I still like the explicitness in what the type name gives though, and I don't like the tediousness a From<u64> introduces (which would be Version::from(0) or worse 0.into())

in most cases you can address this with your method names. Using an example from another comment-

```rust
VersionSelect::From(Version(2))
```

is not really any better than

```rust
VersionSelect::From(2)
```

in this case using `Version` twice is kind of redundant anyway

Reply:
Good point, I'll change it to `pub type Version = u64` 👍🏻